src/main/java/ch/aaap/ca/be/medicalsupplies/MSApplication.java

```diff
@@ -29,83 +38,171 @@ public static void main(String[] args) {
 
     /**
      * Create a model / data structure that combines the input sets.
-     * 
+     *
      * @param genericNameRows
      * @param productRows
      * @return
      */
-    public Object createModel(Set<MSGenericNameRow> genericNameRows, Set<MSProductRow> productRows) {
-        throw new MSException(MSException.DEFAULT_MESSAGE);
+    public Set<Product> createModel(Set<MSGenericNameRow> genericNameRows, Set<MSProductRow> productRows) {
+        Map<String, GenericName> genericNameMap = new HashMap<>();
+        Set<Product> products = new HashSet<>();
+
+        this.genericNames.stream().forEach(msGenericNameRow -> {
+            GenericName genericName = new GenericName();
+            genericName.setName(msGenericNameRow.getName());
+            genericName.setCategories(generateCategories(msGenericNameRow));
+            genericNameMap.put(genericName.getName(), genericName);
+        });
+        this.registry.stream().forEach(msProductRow -> {
+            Producer producer = new Producer(msProductRow.getProducerId()
+                    , msProductRow.getProducerName()
+                    , msProductRow.getProducerAddress());
+            LicenceHolder licenceHolder = new LicenceHolder(msProductRow.getLicenseHolderId()
+                    , msProductRow.getLicenseHolderName()
+                    , msProductRow.getLicenseHolderAddress());
+            Product product = Product.Builder.newInstance()
+                    .id(msProductRow.getId())
+                    .name(msProductRow.getName())
+                    .primaryCategory(new Category(msProductRow.getPrimaryCategory()))
+                    .producer(producer)
+                    .licenceHolder(licenceHolder)
+                    .genericName(genericNameMap.get(msProductRow.getGenericName()))
+                    .build();
+            products.add(product);
+        });
+        return products;
+    }
+
+    private Set<Category> generateCategories(MSGenericNameRow genericNameRow) {
+        Set<Category> categories = new HashSet<>();
+        addCategory(genericNameRow.getCategory1(), categories);
+        addCategory(genericNameRow.getCategory2(), categories);
+        addCategory(genericNameRow.getCategory3(), categories);
+        addCategory(genericNameRow.getCategory4(), categories);
+        return categories;
+    }
+
+    private void addCategory(String categoryName, Set<Category> categories) {
```

Comment:
Our suggestion would be to follow the _pure function_ approach.

```suggestion
    private Category addCategory(String categoryName) {
```